core/src/tasks/event/associateProfile.ts

```diff
@@ -1,30 +1,50 @@
 import { Event } from "../../models/Event";
 import { App } from "../../models/App";
-import { RetryableTask } from "../../classes/tasks/retryableTask";
+import { Task, log } from "actionhero";
+import { CLS } from "../../modules/cls";
+
+const MAX_ATTEMPTS = 2;
+
+export class EventAssociateProfile extends Task {
+  // This Task extends Task rather than CLSTask as we want to be able to view newly created profiles happening in parallel to this task/transaction.
+  // Event creation does have side-effects, so we manually create our own transaction
 
-export class EventAssociateProfile extends RetryableTask {
   constructor() {
     super();
     this.name = "event:associateProfile";
     this.description = "find or create the profile this event is about";
     this.frequency = 0;
     this.queue = "events";
+    this.plugins = ["QueueLock", "Retry"];
     this.inputs = {
       eventGuid: { required: true },
+      count: { required: false },
     };
   }
 
-  async runWithinTransaction(params) {
+  async run(params: { eventGuid: string; count: number }) {
     const { eventGuid } = params;
+    const count = params.count || 0;
     const event = await Event.findByGuid(eventGuid);
 
     const app = await App.findOne({ where: { type: "events" } });
-    // the app might not be in the ready state
-    if (!app) {
-      return;
-    }
-
+    if (!app) return;
     const appOptions = await app.getOptions();
-    await event.associate(appOptions.identifyingPropertyGuid);
+
+    try {
+      await CLS.wrap(async () =>
+        event.associate(appOptions.identifyingPropertyGuid)
+      );
+    } catch (error) {
+      if (count < MAX_ATTEMPTS) {
```

Comment:
why is this different than normal retry behavior?
I think there's a way for the task to have a custom backoff schedule if it's the timing difference.

Reply:
Good idea!

Upside - simpler code. Downside - crazier error logs when a re-enqueue happens. I think I'll still keep the custom CLS wrapper to avoid this:

```
2021-01-12T16:43:13.245Z - info: [ worker ] task success workerId=5 class=event:associateProfile queue=events args={"eventGuid":"evt_e2ea6504-1e0b-440f-b2d2-015b96af0e67"} duration=217
2021-01-12T16:43:13.277Z - info: [ worker ] task success workerId=4 class=event:associateProfile queue=events args={"eventGuid":"evt_7396b367-e682-401c-a4f1-4fa04f70fb7a"} duration=231
2021-01-12T16:43:13.282Z - info: [ worker ] reEnqueue task workerId=3 plugin={"delay":1000,"remainingAttempts":2,"err":{"name":"SequelizeUniqueConstraintError","errors":[{"message":"anonymousId must be unique","type":"unique violation","path":"anonymousId","value":"?-09348f01-49d7-4f5e-9218-efae660ed784","origin":"DB","instance":{"state":"pending","guid":"pro_f3e67ae1-fbb8-4ba7-a419-f979e5e82cb7","anonymousId":"?-09348f01-49d7-4f5e-9218-efae660ed784","updatedAt":"2021-01-12T16:43:13.250Z","createdAt":"2021-01-12T16:43:13.250Z"},"validatorKey":"not_unique","validatorName":null,"validatorArgs":[]}],"fields":{"anonymousId":"?-09348f01-49d7-4f5e-9218-efae660ed784"},"parent":{"length":253,"name":"error","severity":"ERROR","code":"23505","detail":"Key (\"anonymousId\")=(?-09348f01-49d7-4f5e-9218-efae660ed784) already exists.","schema":"public","table":"profiles","constraint":"profiles_anonymous_id","file":"nbtinsert.c","line":"656","routine":"_bt_check_unique","sql":"INSERT INTO \"profiles\" (\"guid\",\"state\",\"anonymousId\",\"createdAt\",\"updatedAt\") VALUES ($1,$2,$3,$4,$5) RETURNING *;","parameters":["pro_f3e67ae1-fbb8-4ba7-a419-f979e5e82cb7","pending","?-09348f01-49d7-4f5e-9218-efae660ed784","2021-01-12 16:43:13.250 +00:00","2021-01-12 16:43:13.250 +00:00"]},"original":{"length":253,"name":"error","severity":"ERROR","code":"23505","detail":"Key (\"anonymousId\")=(?-09348f01-49d7-4f5e-9218-efae660ed784) already exists.","schema":"public","table":"profiles","constraint":"profiles_anonymous_id","file":"nbtinsert.c","line":"656","routine":"_bt_check_unique","sql":"INSERT INTO \"profiles\" (\"guid\",\"state\",\"anonymousId\",\"createdAt\",\"updatedAt\") VALUES ($1,$2,$3,$4,$5) RETURNING *;","parameters":["pro_f3e67ae1-fbb8-4ba7-a419-f979e5e82cb7","pending","?-09348f01-49d7-4f5e-9218-efae660ed784","2021-01-12 16:43:13.250 +00:00","2021-01-12 16:43:13.250 +00:00"]},"sql":"INSERT INTO \"profiles\" (\"guid\",\"state\",\"anonymousId\",\"createdAt\",\"updatedAt\") VALUES ($1,$2,$3,$4,$5) RETURNING *;","error@context":{"_ns_name":"grouparoo-cls","id":588627,"transaction":null}}}
2021-01-12T16:43:13.285Z - info: [ worker ] task success workerId=3 class=event:associateProfile queue=events args={"eventGuid":"evt_29dad04f-87bd-4202-b755-0f69a2b12eb2"} duration=232
2021-01-12T16:43:13.305Z - info: [ worker ] task success workerId=1 class=event:associateProfile queue=events args={"eventGuid":"evt_41cd6be5-4675-4092-b1f0-9aa651bf79d3"} duration=242
```

Reply:
https://github.com/grouparoo/grouparoo/pull/1181